### ISSUE TITLE
Consider using a generic WASM API to read environment properties

### ISSUE THREAD

**Issue by kyessenov** [envoyproxy/envoy-wasm#42]
Existing API is heavily leaning towards using `Struct` as one-fits-all container for passing data from the environment. This forces the WASM clients to include protobuf as a dependency, even for simple things like reading a string from the environment. 

The following API generalizes the existing API to be less protobuf-specific:

Method | Description
--- | ---
Find(StringView…) Handle | Returns a host value using a path of top-level identifier and field selections. This should traverse any nested structures in the environment. The method signature is unrolled up to a certain limit.
Select(Handle, StringView) Handle | Selects a single element in a composite structure (object, map)
Select(Handle, i64) Handle | Selects a single element in a composite list
Contains(Handle, StringView) bool | Test whether a value contains a field or a map contains a key
Contains(Handle, i64) bool | Test whether a value list contains an element
LoadX(Handle) X where X is one of StringView, Struct, I64, F64, Bool | Returns a host value copied into memory module as primitive F64, I64, or StringView, or serialized as a JSON Struct for any composite structure. It is expected the client can either type-check the handle in the client, and the host provides type-checking predicates. Default value is provided if type-checking fails and host monitors the violations.
FindX(StringView…) X | Equivalent to Find followed by LoadX

where

Type | Description
--- | ---
Handle | i32 representing a pointer to a value in the host (composite or primitive); values are typed in the host (number, object, list, map). Errors are values. Null is a value.
StringView | (i32, i32) representing an offset and a length of a string in the module memory
bool | i32

**Comment by kyessenov** [comment]
/cc @TristonianJones as it overlaps with some of CEL activation interface concerns

@PiotrSikora @jplevyak would it be possible to implement a more generic API than the existing barebones `GetMetadata: metadata enum -> protobuf.Struct` at the WASM boundary? I understand that this API could be built on top, but that feels inefficient since it requires plumbing large container structs around.

**Comment by TristonianJones** [comment]
Thanks Kuat, I can see how this lines up with the CEL `Attribute` type and `Resolver` interfaces. 

If I were to model this as Go, I would expect something like the following. Is this roughly right?

```go
v := attr.Variable()
h := Find(v)
for _, elem := range attr.PathElems() {
   h = Select(h, elem)
}
return LoadX(h)
```

**Comment by kyessenov** [comment]
Right. It's just a path traversal with an ability to operate on typed values. Ultimately, it'd be great to run CEL runtimes (cpp or go) on top of that interface, so that we can fallback to dynamic evaluation with the runtime, and rely on optimized code paths (e.g. WASM) that use the interface directly:
```
environment <---- WASM-compatible interface ---> CEL runtime  ------v
                                                                           |-------------------------> compiled code   
```

**Comment by TristonianJones** [comment]
Makes sense. Manage attribute selection within the environment as far as possible, performing native ops on native types if possible. Otherwise, serializing and deferring to the CEL runtime. This seems like the right direction with respect to minimizing evaluation overhead.
